70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+			CHLSLCompiler::SOptions options = {};
+			options.stage = shaderSrc->getStage();
+			if (!(options.stage == IShader::E_SHADER_STAGE::ESS_COMPUTE || options.stage == IShader::E_SHADER_STAGE::ESS_FRAGMENT))
+				options.stage = IShader::E_SHADER_STAGE::ESS_VERTEX;
+			options.targetSpirvVersion = m_device->getPhysicalDevice()->getLimits().spirvVersion;
+			options.spirvOptimizer = nullptr;
+			options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT;
+			options.preprocessorOptions.sourceIdentifier = shaderSrc->getFilepathHint();
+			options.preprocessorOptions.logger = m_logger.get();
+			options.preprocessorOptions.includeFinder = compiler->getDefaultIncludeFinder();
+
+			std::string dxcOptionStr[] = {"-E " + entryPoint};
+			options.dxcOptions = std::span(dxcOptionStr);
+
+			shader = compiler->compileToSPIRV((const char*)shaderSrc->getContent()->getPointer(), options);
+		}
+
+		return m_device->createShader(shader.get());
+	}
+
+	bool createBuffer(smart_refctd_ptr<IGPUBuffer>& buffer, video::IGPUBuffer::SCreationParams& params)
+	{
+		buffer = m_device->createBuffer(std::move(params));
+		if (!buffer)
+			return logFail("Failed to create GPU buffer of size %d!\n", params.size);
+
+		video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = buffer->getMemoryReqs();
+		reqs.memoryTypeBits &= m_physicalDevice->getDeviceLocalMemoryTypeBits();
+
+		auto bufMem = m_device->allocate(reqs, buffer.get());
+		if (!bufMem.isValid())
+			return logFail("Failed to allocate device memory compatible with gpu buffer!\n");
+
+		return true;
+	}
+
+	bool initGraphicsPipeline()
+	{
+		m_renderSemaphore = m_device->createSemaphore(m_submitIx);
+		if (!m_renderSemaphore)
+			return logFail("Failed to create render semaphore!\n");
+			
+		ISwapchain::SCreationParams swapchainParams{
+			.surface = m_surface->getSurface()
+		};
+		if (!swapchainParams.deduceFormat(m_physicalDevice))
+			return logFail("Could not choose a surface format for the swapchain!\n");
+
+		const static IGPURenderpass::SCreationParams::SSubpassDependency dependencies[] = {
+			{
+				.srcSubpass = IGPURenderpass::SCreationParams::SSubpassDependency::External,
+				.dstSubpass = 0,
+				.memoryBarrier = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::LATE_FRAGMENT_TESTS_BIT,
+					.srcAccessMask = ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::EARLY_FRAGMENT_TESTS_BIT,
+					.dstAccessMask = ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_WRITE_BIT | ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_READ_BIT
+				}
+			},
+			// color from ATTACHMENT_OPTIMAL to PRESENT_SRC
+			{
+				.srcSubpass = 0,
+				.dstSubpass = IGPURenderpass::SCreationParams::SSubpassDependency::External,
+				.memoryBarrier = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COLOR_ATTACHMENT_OUTPUT_BIT,
+					.srcAccessMask = ACCESS_FLAGS::COLOR_ATTACHMENT_WRITE_BIT
+				}
+			},
```

Comment:
yes present has no DST stage, but you also need a barrier TO (dst) the compute shader of the next submit that will modify the particle SSBO and grid Images you read from during rendering